Source/DafnyCore/AST/IndentationFormatter.cs

```diff
@@ -0,0 +1,2393 @@
+    };
+    foreach (var include in program.DefaultModuleDef.Includes) {
+      if (include.OwnedTokens.Any()) {
+        indentationFormatter.SetOpeningIndentedRegion(include.OwnedTokens.First(), 0);
+      }
+    }
+
+    foreach (var decl in program.DefaultModuleDef.TopLevelDecls) {
+      indentationFormatter.SetDeclIndentation(decl, 0);
+    }
+    foreach (var decl in program.DefaultModuleDef.PrefixNamedModules) {
+      indentationFormatter.SetDeclIndentation(decl.Item2, 0);
+    }
+
+    return indentationFormatter;
+  }
+
+  protected IEnumerable<Expression> SubExpressions(Expression expr) {
+    return expr is ConcreteSyntaxExpression concreteSyntaxExpression
+      ? concreteSyntaxExpression.PreResolveSubExpressions
+      : expr.SubExpressions;
+  }
+
+
+  protected override bool VisitOneExpr(Expression expr, ref int unusedIndent) {
+    if (expr == null) {
+      return false;
+    }
+    var firstToken = expr.StartToken;
+    var indent = GetIndentBefore(firstToken);
+    switch (expr) {
+      case ChainingExpression chainingExpr:
+        return SetChainingExprIndent(indent, chainingExpr);
+      case BinaryExpr binaryExpr:
+        return SetBinaryExprIndent(indent, binaryExpr);
+      case LetOrFailExpr:
+      case LetExpr:
+        return SetIndentVarDeclStmt(indent, expr.OwnedTokens, expr is LetOrFailExpr { Lhs: null }, true);
+      case ITEExpr iteExpr:
+        return SetIndentITE(indent, iteExpr);
+      case NestedMatchExpr nestedMatchExpr:
+        var i = unusedIndent;
+        return SetIndentCases(indent, expr.OwnedTokens.Concat(nestedMatchExpr.Cases.SelectMany(oneCase => oneCase.OwnedTokens)).OrderBy(token => token.pos), () => {
+          foreach (var e in SubExpressions(expr)) {
+            Visit(e, i);
+          }
+        });
+      case LambdaExpr lambaExpression:
+        return SetIndentLambdaExpr(indent, lambaExpression);
+      case ComprehensionExpr:
+        return SetIndentComprehensionExpr(indent, expr.OwnedTokens);
+      case UnaryExpr:
+      case OldExpr:
+      case UnchangedExpr:
+      case ParensExpression:
+      case SeqDisplayExpr:
+      case MapDisplayExpr:
+      case SetDisplayExpr:
+        return SetIndentParensExpression(indent, expr.OwnedTokens);
+      case ApplySuffix applySuffix: {
+          var reindent = ReduceBlockiness ? indent
+            : GetNewTokenVisualIndent(applySuffix.StartToken, indent);
+          return SetIndentParensExpression(reindent, expr.OwnedTokens);
+        }
+      case StmtExpr stmtExpr: {
+          return SetIndentStmtExpr(indent, stmtExpr);
+        }
+    }
+
+    return true;
+  }
+
+  protected override bool VisitOneStmt(Statement stmt, ref int unusedIndent) {
+    if (stmt == null) {
+      return false;
+    }
+    var firstToken = stmt.StartToken;
+    var indent = GetIndentBefore(firstToken);
+    // Every function returns if traverse needs to occur (true) or if it already happened (false) 
+    switch (stmt) {
```

Comment:
This file is very large, making it harder to comment on in GitHub and slowing than my IDE. Have you considered moving some of these methods to the AST nodes that they apply to?

You might even have interface `ICanFormat` with a method `SetIndent(IndentationFormatter formatter` that these AST nodes can implement, so you don't need a switch like this.

Reply:
That's a good idea. I liked it as well when it was centralized but it's also nice to delegate this to nodes themselves. Let me do it.

Reply:
Amazing, thanks a lot!